### PR TITLE
Topology tolerance [AVD-1723]

### PR DESCRIPTION
**Pull request by pp-mo**
### Description
Allow "wrong" topology-dimension on a UGRID mesh, deduce "correct" value from attached connectivites.

This is just one of the pending "tolerant loading" ideas which we listed elsewhere (AVD-1723).
But this one will help with the forthcoming internal release demo.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

File: lib/iris/experimental/ugrid/__init__.py
Comment (stephenworsley) [inline lib/iris/experimental/ugrid/__init__.py:3670]: Would it not make more sense to set `topology_dimension = expected_topology_dimension` outside of this "if" statement and have this logic only handle warnings? It seems to me that there's currently no case where `topology_dimension` does not end up being equal to `expected_topology_dimension`.

**Review by stephenworsley — CHANGES_REQUESTED** [review]
Looks good, just a few comments to address.

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (pp-mo) [reply lib/iris/tests/integration/experimental/test_ugrid_load.py:176]: :+1:

File: lib/iris/experimental/ugrid/__init__.py
Comment (pp-mo) [reply lib/iris/experimental/ugrid/__init__.py:3670]: Ok, well put.
I will try + make this clearer ... :+1:

**Comment by pp-mo** [comment]
Thanks @stephenworsley  
I hope the latest commit addresses your sugesstions
Please re-review !

File: lib/iris/experimental/ugrid/__init__.py
Comment (trexfeathers) [inline lib/iris/experimental/ugrid/__init__.py:20]: For the ugrid module, @bjlittle and I have been so far been deliberately avoiding `warnings.warn()` - **in favour of `logger.debug()`** - on the understanding that this pattern is being gradually propagated throughout Iris to make the user experience less 'noisy'.

Comment (pp-mo) [reply lib/iris/experimental/ugrid/__init__.py:20]: Well, I considered that.  But I was trying to be consistent with other things in the netcdf loading code -- in the Pyke arena, mostly.

Comment (trexfeathers) [reply lib/iris/experimental/ugrid/__init__.py:20]: I see. Should I therefore raise a separate pull request to convert the existing code in this module to use `warnings.warn()` instead of `logger.debug()` ([e.g. L3336](https://github.com/SciTools/iris/blob/6e329d28fe40cf07735d95e7ad3730851216570d/lib/iris/experimental/ugrid/__init__.py#L3336))?

**Review by stephenworsley — APPROVED** [review]
Looks good, I think this is good to merge.

File: lib/iris/experimental/ugrid/__init__.py
Comment (pp-mo) [reply lib/iris/experimental/ugrid/__init__.py:20]: Well, I'd expect it to be more likely the other way about : we should eventually converge on logging for everything.
I was just hoping to be consistent with other parts of the loading code, for now.
But now I realise that I don't know enough about what has been done elsewhere recently.

I also see you are also using logger.debug rather than logger.warning, which I don't really get -- aren't these warning-level problems?  Does this represent a new intention to be silent unless the user specifically asks about it ?
I was expecting to see a message by default, but make it easy to hide.

Comment (trexfeathers) [reply lib/iris/experimental/ugrid/__init__.py:20]: >Does this represent a new intention to be silent unless the user specifically asks about it ?

Yeah that's where this is coming from. To be fair to you I was also reticent about this when talking to Bill, but it's what has already been adopted during the cube arithmetic work.

This deserves a team conversation soon if examples have already gone into `master` (cube arithmetic), but we're not all on the same page.

I don't know what to do here. Happy to leave it to @pp-mo and @stephenworsley to judge.

Comment (pp-mo) [reply lib/iris/experimental/ugrid/__init__.py:20]: Here's my plan : 
  * use logging in place of warnings
  * but use logging.warning instead of logging.debug/info in this case.
     * looking at other usages, I think it is perfectly consistent to use different levels for different cases (though I do still think that some of the existing levels are wrong)
  * it also means developing a means of testing for a logged message [in place of a warning](https://github.com/SciTools/iris/blob/6e329d28fe40cf07735d95e7ad3730851216570d/lib/iris/tests/integration/experimental/test_ugrid_load.py#L160).  
     So, I'll put this PR into draft + re-enable it when I've fixed that ...

Comment (trexfeathers) [reply lib/iris/experimental/ugrid/__init__.py:20]: >though I do still think that some of the existing levels are wrong

Agreed

> it also means developing a means of testing for a logged message [in place of a warning](https://github.com/SciTools/iris/blob/6e329d28fe40cf07735d95e7ad3730851216570d/lib/iris/tests/integration/experimental/test_ugrid_load.py#L160).

There are plenty of examples testing for logged `DEBUG` messages in the existing experimental.ugrid unit tests 👍

Comment (pp-mo) [reply lib/iris/experimental/ugrid/__init__.py:20]: > examples testing for logged DEBUG messages

oh, thanks for the hint @trexfeathers !

**Comment by pp-mo** [comment]
Back, with requested changes to warnings/logging.

@trexfeathers and/or @stephenworsley : Please re-review !

**Comment by pp-mo** [comment]
Thanks @trexfeathers for pointing out the existing functionality for logging tests.
I was **_very_** close to committing effort to writing that myself !

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/test_ugrid_load.py:166]: Deprecated - use `assertRegex()` instead.

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (trexfeathers) [inline lib/iris/tests/integration/experimental/test_ugrid_load.py:181]: Deprecated - use `assertRegex()` instead.

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Thanks @pp-mo! Couple of little things to address.

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (pp-mo) [reply lib/iris/tests/integration/experimental/test_ugrid_load.py:160]: Yes, but I decided to fix this a different way (see latest commit) ...

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (pp-mo) [reply lib/iris/tests/integration/experimental/test_ugrid_load.py:178]: :+1:

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (pp-mo) [reply lib/iris/tests/integration/experimental/test_ugrid_load.py]: :+1:

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (trexfeathers) [reply lib/iris/tests/integration/experimental/test_ugrid_load.py:160]: Nice. Is `level="DEBUG"` even necessary at all? If not, its presence is just an opportunity to confuse future devs.

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (trexfeathers) [reply lib/iris/tests/integration/experimental/test_ugrid_load.py:178]: https://github.com/SciTools/iris/pull/4099#discussion_r617373784

File: lib/iris/tests/integration/experimental/test_ugrid_load.py
Comment (trexfeathers) [reply lib/iris/tests/integration/experimental/test_ugrid_load.py:160]: I've tried this myself - it can definitely be removed without a problem. You OK to do that?

Comment (pp-mo) [reply lib/iris/tests/integration/experimental/test_ugrid_load.py:160]: Sure, good idea !
I see from tracing the code that effectively the default is INFO.  Shame it's not documented :wink:

**Comment by pp-mo** [comment]
Thanks for suggestion.  Done that, I hope ?
BTW I also moved the support vars for the synthetic data calls inside the logging context block, which I think reads better.

Please re-review @trexfeathers

**Review by trexfeathers — APPROVED** [review]

**Comment by trexfeathers** [comment]
Merging despite the link check failures. Once #4104 is in, we'll do a merge-back to get that fix into [mesh-data-model](https://github.com/SciTools/iris/tree/mesh-data-model).

**Comment by pp-mo** [comment]
Many thanks, significantly better for your suggestions @stephenworsley @trexfeathers !
💐

**Pull request merged**
